File: kernel/src/term/font.rs
use crate::term::Color;

pub struct Font {
    chars: [[[u8; Self::CHAR_WIDTH]; Self::CHAR_HEIGHT]; 128],
}

impl Font {
    const CHAR_WIDTH: usize = 11;
    const CHAR_HEIGHT: usize = 20;

    pub fn width_pixels(&self) -> usize {
        Self::CHAR_WIDTH
    }
    pub fn height_pixels(&self) -> usize {
        Self::CHAR_HEIGHT
    }

    fn new() -> Self {
        let buffer: &mut [u8] = {
            static mut BUFFER: [u8; 1 << 19] = [0; 1 << 19];
            unsafe { &mut BUFFER }
        };

        let bytes = include_bytes!("../../../font.png");

        let mut png = minipng::decode_png(bytes, buffer).expect("Failed to decode font PNG");
        let columns = 20;

        png.convert_to_rgba8bpc()
            .expect("Failed to convert font PNG to RGBA8");

        let mut chars = [[[0; Self::CHAR_WIDTH]; Self::CHAR_HEIGHT]; 128];

        for c in ' '..'~' {
            let char_index = (c as u32 - ' ' as u32) as usize;
            let tile_position = (
                (char_index % columns) * Self::CHAR_WIDTH,
                (char_index / columns) * Self::CHAR_HEIGHT,
            );

            for y in 0..Self::CHAR_HEIGHT {
                for x in 0..Self::CHAR_WIDTH {
                    let pixel_index = (tile_position.1 + y) * (columns * Self::CHAR_WIDTH)
                        + (tile_position.0 + x);
                    let alpha = png.pixels()[pixel_index as usize * 4 + 3];

                    chars[char_index][y][x] = alpha;
                }
            }
        }

        Self { chars }
    }

    pub fn blit_char(&self, c: char, ptr: *mut Color, row_pitch: usize, fg: Color, bg: Color) {
        let bitmap = self
            .chars
            .get(c as usize - ' ' as usize)
            .expect("Character not found in font");

        for row in 0..Self::CHAR_HEIGHT {
            for col in 0..Self::CHAR_WIDTH {
                let ptr = unsafe { ptr.add(row_pitch * row + col) };

                let alpha = bitmap[row][col];

                if alpha > 0 {
                    unsafe {
                        ptr.write_volatile(fg.blend(bg, alpha));
                    }
                }
            }
        }
    }
}

pub static FONT_1: spin::Lazy<Font> = spin::Lazy::new(Font::new);

File: kernel/src/term/font.zig
const std = @import("std");
const zigimg = @import("zigimg");

const PNG = @embedFile("font.png");

pub const Font = struct {
    pub const CHAR_WIDTH: usize = 11;
    pub const CHAR_HEIGHT: usize = 20;
    pub const CHAR_COUNT: usize = 128;

    chars: [CHAR_COUNT][CHAR_HEIGHT][CHAR_WIDTH]u8,

    pub fn init() !Font {
        var buffer: [0x100_000]u8 = undefined;
        var allocator = std.heap.FixedBufferAllocator.init(&buffer);

        var image = try zigimg.Image.fromMemory(allocator.allocator(), PNG);
        defer image.deinit();

        try image.convert(.grayscale8Alpha);

        var font = Font{ .chars = undefined };

        for (' '..'~') |c| {
            const char_index = @as(usize, @intCast(c)) - @as(usize, @intCast(' '));
            const char_x = (c - ' ') % 20 * CHAR_WIDTH;
            const char_y = (c - ' ') / 20 * CHAR_HEIGHT;

            for (0..CHAR_HEIGHT) |y| {
                for (0..CHAR_WIDTH) |x| {
                    font.chars[char_index][y][x] = image.pixels.grayscale8Alpha.ptr[(char_y + y) * image.width + (char_x + x)].alpha;
                }
            }
        }

        return font;
    }

    pub fn getChar(self: Font, c: u8) ?[CHAR_HEIGHT][CHAR_WIDTH]u8 {
        if (c < ' ' or c >= '~') {
            return null;
        }
        const char_index = @as(usize, @intCast(c)) - @as(usize, @intCast(' '));
        return self.chars[char_index];
    }
};
